ArduCopter/mode_smart_rtl.cpp

```diff
@@ -92,42 +92,45 @@ void ModeSmartRTL::path_follow_run()
 
     // if we are close to current target point, switch the next point to be our target.
     if (wp_nav->reached_wp_destination()) {
-        Vector3f dest_NED;
-        // this pop_point can fail if the IO task currently has the
-        // path semaphore.
-        if (g2.smart_rtl.pop_point(dest_NED)) {
-            path_follow_last_pop_fail_ms = 0;
-            if (g2.smart_rtl.get_num_points() == 0) {
-                // this is the very last point, add 2m to the target alt and move to pre-land state
-                dest_NED.z -= 2.0f;
+        HAL_Semaphore &sem = g2.smart_rtl.path_semaphore();
```

Comment:
I thought we always used a WITH_SEMAPHORE macro instead of writing the take and give directly.

Reply:
In this case we're not taking this  semaphore in a blocking fashion - we're using take_nonblocking.  As we're in the main thread and the path-tidying thread can take the path sempahore for prolonged periods of time so we don't want to block.

Doing it this way does mean we have to be extra-special-careful to release the semaphore, of course.

Note that this is pre-existing.  I've just moved the "take" of the semaphore to be out around all of the work the main thread is doing to ensure consistency of the path count with the pop/peek operations.